src/main/java/org/truffleruby/core/string/StringNodes.java

```diff
@@ -365,16 +365,40 @@ protected boolean isRubyString(DynamicObject object) {
     public abstract static class CompareNode extends CoreMethodArrayArgumentsNode {
 
         @Specialization(guards = "isRubyString(b)")
-        public int compare(DynamicObject a, DynamicObject b) {
+        public int compare(DynamicObject a, DynamicObject b,
+                           @Cached("createBinaryProfile()") ConditionProfile sameRopeProfile,
+                           @Cached("createBinaryProfile()") ConditionProfile equalSubsequenceProfile,
+                           @Cached("createBinaryProfile()") ConditionProfile equalProfile,
+                           @Cached("createBinaryProfile()") ConditionProfile notComparableProfile,
+                           @Cached("create()") RopeNodes.BytesNode firstBytesNode,
+                           @Cached("create()") RopeNodes.BytesNode secondBytesNode) {
             // Taken from org.jruby.RubyString#op_cmp
 
             final Rope firstRope = rope(a);
             final Rope secondRope = rope(b);
 
-            final int ret = RopeOperations.cmp(firstRope, secondRope);
+            if (sameRopeProfile.profile(firstRope == secondRope)) {
+                return 0;
+            }
+
+            final int size = firstRope.byteLength();
+            final int len =  Math.min(size, secondRope.byteLength());
 
-            if ((ret == 0) && !RopeOperations.areComparable(firstRope, secondRope)) {
-                return firstRope.getEncoding().getIndex() > secondRope.getEncoding().getIndex() ? 1 : -1;
+            final byte[] bytes = firstBytesNode.execute(firstRope);
+            final byte[] otherBytes = secondBytesNode.execute(secondRope);
+
+            final int ret;
+            final int cmp = ArrayUtils.memcmp(bytes, 0, otherBytes, 0, len);
+            if (equalSubsequenceProfile.profile(cmp == 0)) {
+                ret = size == secondRope.byteLength() ? 0 : size == len ? -1 : 1;
```

Comment:
I would also profile this condition, so it's clear to PE whether 0 can be the result or not.